# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

导入数据分析所需要的库，并通过pandas的read_csv函数，将原始数据文件“e_commerce.csv”里面的数据内容，解析为DataFrame，并且赋值给变量original_data

In [1]:
import pandas as pd

In [3]:
original_data = pd.read_csv("e_commerce.csv")

In [4]:
original_data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


## 评估数据

对上一部分建立的original_data这个DataFrame所包含的数据进行评估。
评估主要从两个方面进行：结构和内容，即整齐度和干净度。
数据的结构性问题指不符合“每行一个观察值，每列一个变量，每个单元格是一个值”这三个标准，数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

### 评估数据整齐度

In [10]:
original_data.sample(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
414932,572499,23557,LANDMARK FRAME BAKER STREET,3,10/24/2011 14:10,12.50,15786.0,United Kingdom
484444,577531,23354,6 GIFT TAGS 50'S CHRISTMAS,9,11/20/2011 13:58,1.63,NaN,United Kingdom
484128,577521,21287,SCENTED VELVET LOUNGE CANDLE,12,11/20/2011 13:23,0.38,16411.0,United Kingdom
50717,540551,90014B,GOLD M PEARL ORBIT NECKLACE,1,1/10/2011 9:43,10.62,NaN,United Kingdom
403393,571564,23355,HOT WATER BOTTLE KEEP CALM,4,10/18/2011 10:16,4.95,16019.0,United Kingdom
113883,545999,22948,METAL DECORATION NAUGHTY CHILDREN,24,3/8/2011 14:44,0.85,13089.0,United Kingdom
404317,571671,85199S,SMALL HANGING IVORY/RED WOOD BIRD,12,10/18/2011 13:15,0.42,16940.0,United Kingdom
257189,559518,20828,GLITTER BUTTERFLY CLIPS,1,7/8/2011 16:11,4.96,NaN,United Kingdom
44529,540179,22191,IVORY DINER WALL CLOCK,4,1/5/2011 12:54,8.50,16033.0,United Kingdom
172171,551520,85049A,TRADITIONAL CHRISTMAS RIBBONS,3,5/1/2011 11:55,1.25,13301.0,United Kingdom


数据符合“每行是一个观察值，每列是一个变量，每个单元格是一个值”的标准，每行是关于某商品的一次交易，每列是交易相关的各个变量，因此不存在结构性问题

### 评估数据干净度

In [11]:
original_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  str    
 1   StockCode    541909 non-null  str    
 2   Description  540455 non-null  str    
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  str    
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 33.1 MB


数据共有541909条观察值，而`Description`、`CustomerID`具有缺失值。
`InvoiceDate`的数据类型应该为日期，`CustomerID`的数据类型应该为字符串，应该进行数据格式转化。

#### 评估缺失数据

In [13]:
original_data[original_data["Description"].isnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,12/1/2010 14:32,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,12/1/2010 14:34,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,12/7/2011 18:26,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,12/7/2011 18:31,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,12/7/2011 18:35,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,12/8/2011 10:33,0.0,NaN,United Kingdom


In [16]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


筛选出来结果数量为0条，说明缺失`Description`的数据同时也不具备有效的`UnitPrice`值。
`Description`表示产品名称，`UnitPrice`表示产品单价，都是后续进行商品交易分析的重要变量。如果他们同时缺失/无效，我们认为数据无法提供有效含义，因此可以后续删除。

In [17]:
original_data[original_data["CustomerID"].isnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.00,NaN,United Kingdom
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,12/1/2010 14:32,2.51,NaN,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,12/1/2010 14:32,2.51,NaN,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,12/1/2010 14:32,0.85,NaN,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,12/1/2010 14:32,1.66,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
541536,581498,85099B,JUMBO BAG RED RETROSPOT,5,12/9/2011 10:26,4.13,NaN,United Kingdom
541537,581498,85099C,JUMBO BAG BAROQUE BLACK WHITE,4,12/9/2011 10:26,4.13,NaN,United Kingdom
541538,581498,85150,LADIES & GENTLEMEN METAL SIGN,1,12/9/2011 10:26,4.96,NaN,United Kingdom
541539,581498,85174,S/4 CACTI CANDLES,1,12/9/2011 10:26,10.79,NaN,United Kingdom


`CustomerID`表示客户编号，不是分析畅销商品的必要变量，所以这些数据仍然有效，因此保留此变量为空的观察值。

#### 评估重复数据

虽然`InvoiceNo`、`StockCode`和`CustomerID`都是唯一标识符，但一次交易可能包含多件商品，所以`InvoiceNo`可以重复；不同交易可以包含同一件商品，所以`StockCode`可以重复；顾客可以多次交易或者下单多个商品，所以`CustomerID`可以重复。
因此，针对此数据我们无需评估重复数据。

#### 评估不一致数据

`Country`变量中，我们要查看是否存在多个不同值指代同一个国家的情况。

In [26]:
original_data["Country"].value_counts().sort_index()

Country
Australia                 1259
Austria                    401
Bahrain                     19
Belgium                   2069
Brazil                      32
Canada                     151
Channel Islands            758
China                      288
Cyprus                     622
Czech Republic              30
Denmark                    389
EIRE                      8196
European Community          61
Finland                    695
France                    8557
Germany                   9495
Greece                     146
Iceland                    182
Israel                     297
Italy                      803
Japan                      358
Lebanon                     45
Lithuania                   35
Malta                      127
Netherlands               2371
Norway                    1086
Poland                     341
Portugal                  1519
RSA                         58
Saudi Arabia                10
Singapore                  229
Spain                     2533


`USA`、`United States`都在表示美国，`UK`、`U.K.`、`United Kingdom`都在表示英国，应该对其统一只保留一个指代值。

#### 评估无效或错误数据

In [27]:
original_data.describe()

,Quantity,UnitPrice,CustomerID
count,541909.000000,541909.000000,406829.000000
mean,9.552250,4.611114,15287.690570
std,218.081158,96.759853,1713.600303
min,-80995.000000,-11062.060000,12346.000000
25%,1.000000,1.250000,13953.000000
50%,3.000000,2.080000,15152.000000
75%,10.000000,4.130000,16791.000000
max,80995.000000,38970.000000,18287.000000


从输出结果来看，`Quantity`、`UnitPrice`存在负数。
应该筛选出来进一步评估其含义。

In [30]:
original_data[original_data["Quantity"] < 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
433987,C574019,22738,RIBBON REEL SNOWY VILLAGE,-120,11/2/2011 12:05,1.45,12415.0,Australia
210659,C555288,23079,TOADSTOOL BEDSIDE LIGHT,-2,6/2/2011 9:57,8.95,12431.0,Australia
210658,C555288,84945,MULTI COLOUR SILVER T-LIGHT HOLDER,-6,6/2/2011 9:57,0.85,12431.0,Australia
270065,C560540,23244,ROUND STORAGE TIN VINTAGE LEAF,-1,7/19/2011 12:26,1.95,12415.0,Australia
270066,C560540,23243,SET OF TEA COFFEE SUGAR TINS PANTRY,-1,7/19/2011 12:26,4.95,12415.0,Australia
...,...,...,...,...,...,...,...,...
536923,C581229,22712,CARD DOLLY GIRL,-12,12/8/2011 10:14,0.42,12558.0,United States
536916,C581229,85099B,JUMBO BAG RED RETROSPOT,-10,12/8/2011 10:14,2.08,12558.0,United States
536925,C581229,21508,VINTAGE KID DOLLY CARD,-12,12/8/2011 10:14,0.42,12558.0,United States
536921,C581229,23209,LUNCH BAG VINTAGE DOILY,-20,12/8/2011 10:14,1.65,12558.0,United States


In [32]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C")]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
12069,537363,84569B,NaN,-20,12/6/2010 12:35,0.0,NaN,United Kingdom
12070,537364,21027,NaN,-70,12/6/2010 12:36,0.0,NaN,United Kingdom
12934,537401,84898F,NaN,-24,12/6/2010 14:37,0.0,NaN,United Kingdom
13219,537427,84968B,NaN,-21,12/6/2010 15:37,0.0,NaN,United Kingdom
13217,537425,84968F,check,-20,12/6/2010 15:35,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
510117,579429,20711,lost in space,-98,11/29/2011 13:04,0.0,NaN,United Kingdom
521333,580359,20775,NaN,-35,12/2/2011 16:11,0.0,NaN,United Kingdom
524472,580561,22043,dotcom,-9,12/5/2011 10:25,0.0,NaN,United Kingdom
524369,580546,23406,CHECK,-21,12/5/2011 9:27,0.0,NaN,United Kingdom


In [35]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C") & (original_data["UnitPrice"] != 0)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


猜想得到验证，当`Quantity`变量为负数的时候，观察值满足以下条件之一：  
- `InvoiceNo`以C开头，表示订单被取消
- `UnitPrice`为0，说明单价为0英镑
   
这些交易数据都不是有效成交数据，不在后续分析范围之内，所以我们将在数据清理步骤里将`Quantity`为负数的观察值删除。

In [36]:
original_data[original_data["UnitPrice"] < 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,A563186,B,Adjust bad debt,1,8/12/2011 14:51,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,8/12/2011 14:52,-11062.06,NaN,United Kingdom


`UnitPrice`为负数的都是坏账调整，不属于实际商品交易数据，在数据清理步骤中也将其删除。

## 清洗数据

根据前面评估部分得到的结论，需要进行的数据清理包括：  
- `InvoiceDate`的数据类型转化为日期
- `CustomerID`的数据类型转化为字符串
- `Description`变量缺失的观察值删除
- 把`Country`变量值"USA"替换为"United States"
- 把`Country`变量值"UK"、"U.K."替换成"United Kingdom"
- 将`Quantity`为负数的观察值删除
- 将`UnitPrice`为负数的观察值删除

In [39]:
cleaned_data = original_data.copy()
cleaned_data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
228227,556917,22085,PAPER CHAIN KIT SKULLS,40,6/15/2011 13:37,2.55,12415.0,Australia
269595,560491,21843,RED RETROSPOT CAKE STAND,1,7/19/2011 10:51,10.95,12415.0,Australia
269596,560491,20979,36 PENCILS TUBE RED RETROSPOT,1,7/19/2011 10:51,1.25,12415.0,Australia
269597,560491,23245,SET OF 3 REGENCY CAKE TINS,1,7/19/2011 10:51,4.95,12415.0,Australia
228262,556917,22704,WRAP RED APPLES,400,6/15/2011 13:37,0.34,12415.0,Australia


`InvoiceDate`的数据类型转化为日期

In [43]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

228227   2011-06-15 13:37:00
269595   2011-07-19 10:51:00
269596   2011-07-19 10:51:00
269597   2011-07-19 10:51:00
228262   2011-06-15 13:37:00
                 ...        
308810   2011-08-22 13:32:00
308809   2011-08-22 13:32:00
282749   2011-07-28 16:06:00
498757   2011-11-24 14:55:00
282857   2011-07-28 16:21:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[us]

 `CustomerID`的数据类型转化为字符串

In [44]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"]

228227    12415.0
269595    12415.0
269596    12415.0
269597    12415.0
228262    12415.0
           ...   
308810    14265.0
308809    14265.0
282749    12743.0
498757        NaN
282857    12743.0
Name: CustomerID, Length: 541909, dtype: str

 `CustomerID`的变量值结尾“.0”去掉

In [48]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0,-2)
cleaned_data["CustomerID"]

228227    12415
269595    12415
269596    12415
269597    12415
228262    12415
          ...  
308810    14265
308809    14265
282749    12743
498757      NaN
282857    12743
Name: CustomerID, Length: 541909, dtype: str

`Description`变量缺失的观察值删除

In [50]:
cleaned_data.dropna(subset=["Description"],inplace=True)

In [52]:
cleaned_data["Description"].isnull().sum()

np.int64(0)

把`Country`变量值"USA"替换为"United States"

In [57]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA":"United States"})

In [58]:
len(cleaned_data[cleaned_data["Country"] == "USA"])

0

把`Country`变量值"UK"、"U.K."替换成"United Kingdom"

In [60]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK":"United Kingdom","U.K.":"United Kingdom"})

In [61]:
len(cleaned_data[cleaned_data["Country"] == "UK"])

0

In [62]:
len(cleaned_data[cleaned_data["Country"] == "U.K."])

0

将`Quantity`为负数的观察值删除

In [63]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]

In [64]:
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

将`UnitPrice`为负数的观察值删除

In [65]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]

In [66]:
len(cleaned_data[cleaned_data["UnitPrice"] < 0])

0

## 保存清理后的数据

In [67]:
cleaned_data.to_csv("e_commerce_cleaned.csv",index = False)